File: src/系统/03．技能系统/00．技能模板+函数/00．技能模板/04．主动技能流程模板/00．流程生命周期.lua
local ____lualib = require("lualib_bundle")
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____on_4E3B_52A8_6280_80FD_6D41_7A0B_5355_4F4D_6B7B_4EA1, _____6D3B_8DC3_6D41_7A0B_5217_8868
function ____on_4E3B_52A8_6280_80FD_6D41_7A0B_5355_4F4D_6B7B_4EA1(dyingUnit)
    do
        local i = #_____6D3B_8DC3_6D41_7A0B_5217_8868 - 1
        while i >= 0 do
            local _____5B9E_4F8B = _____6D3B_8DC3_6D41_7A0B_5217_8868[i + 1]
            if _____5B9E_4F8B ~= nil then
                _____5B9E_4F8B["处理单位死亡"](_____5B9E_4F8B, dyingUnit)
            end
            i = i - 1
        end
    end
end
local jass = require("jass.common")
local IsUnitType = jass.IsUnitType
local UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD
local ____require_result_0 = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心")
local registerDeathListener = ____require_result_0.registerDeathListener
_____6D3B_8DC3_6D41_7A0B_5217_8868 = {}
local _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C = false
local function _____5355_4F4D_6709_6548(unit)
    return unit ~= nil and unit ~= 0 and IsUnitType(unit, UNIT_TYPE_DEAD) ~= true
end
local function _____79FB_9664_6D3B_8DC3_6D41_7A0B(_____5B9E_4F8B)
    do
        local i = #_____6D3B_8DC3_6D41_7A0B_5217_8868 - 1
        while i >= 0 do
            if _____6D3B_8DC3_6D41_7A0B_5217_8868[i + 1] == _____5B9E_4F8B then
                __TS__ArraySplice(_____6D3B_8DC3_6D41_7A0B_5217_8868, i, 1)
                break
            end
            i = i - 1
        end
    end
end
local function _____786E_4FDD_6B7B_4EA1_76D1_542C()
    if _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C then
        return
    end
    _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C = true
    registerDeathListener(____on_4E3B_52A8_6280_80FD_6D41_7A0B_5355_4F4D_6B7B_4EA1)
end
local function ____on_4E3B_52A8_6280_80FD_6D41_7A0B_6E05_7406(variable)
    local _____5B9E_4F8B = variable
    if _____5B9E_4F8B ~= nil then
        _____5B9E_4F8B["停止"](_____5B9E_4F8B, "清理")
    end
end
local _____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0 = __TS__Class()
_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.name = "主动技能流程生命周期实现"
function _____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.prototype.____constructor(self, _____53C2_6570)
    self["已结束值"] = false
    self["参数"] = _____53C2_6570
    local ____opt_1 = _____53C2_6570["清理"]
    if ____opt_1 and ____opt_1["已清理"](____opt_1) then
        self["结束内部"](self, "清理", false)
        return
    end
    if _____53C2_6570["施法者死亡时取消"] ~= false and not _____5355_4F4D_6709_6548(_____53C2_6570["施法者"]) then
        self["结束内部"](self, "死亡", false)
        return
    end
    if _____53C2_6570["目标死亡时取消"] == true and _____53C2_6570["目标"] ~= nil and _____53C2_6570["目标"] ~= 0 and not _____5355_4F4D_6709_6548(_____53C2_6570["目标"]) then
        self["结束内部"](self, "目标失效", false)
        return
    end
    _____6D3B_8DC3_6D41_7A0B_5217_8868[#_____6D3B_8DC3_6D41_7A0B_5217_8868 + 1] = self
    _____786E_4FDD_6B7B_4EA1_76D1_542C()
    if _____53C2_6570["清理"] ~= nil then
        local ____self_3 = _____53C2_6570["清理"]
        ____self_3["登记清理"](____self_3, _____53C2_6570["名称"] .. "-停止流程", ____on_4E3B_52A8_6280_80FD_6D41_7A0B_6E05_7406, self)
    end
end
_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.prototype["停止"] = function(self, _____539F_56E0)
    if _____539F_56E0 == nil then
        _____539F_56E0 = "中断"
    end
    return self["结束内部"](self, _____539F_56E0, true)
end
_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.prototype["结束"] = function(self, _____539F_56E0)
    return self["结束内部"](self, _____539F_56E0, false)
end
_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.prototype["完成"] = function(self)
    return self["结束内部"](self, "完成", false)
end
_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.prototype["是否结束"] = function(self)
    return self["已结束值"]
end
_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.prototype["读取结束原因"] = function(self)
    return self["结束原因"]
end
_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.prototype["处理单位死亡"] = function(self, dyingUnit)
    if self["已结束值"] then
        return
    end
    if self["参数"]["施法者死亡时取消"] ~= false and dyingUnit == self["参数"]["施法者"] then
        self["停止"](self, "死亡")
        return
    end
    if self["参数"]["目标死亡时取消"] == true and dyingUnit == self["参数"]["目标"] then
        self["停止"](self, "目标失效")
    end
end
_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0.prototype["结束内部"] = function(self, _____539F_56E0, _____8C03_7528_505C_6B62_56DE_8C03)
    if self["已结束值"] then
        return false
    end
    self["已结束值"] = true
    self["结束原因"] = _____539F_56E0
    _____79FB_9664_6D3B_8DC3_6D41_7A0B(self)
    if _____8C03_7528_505C_6B62_56DE_8C03 and self["参数"]["on停止"] ~= nil then
        self["参数"]["on停止"](_____539F_56E0, self["参数"]["变量"])
    end
    if self["参数"]["on结束"] ~= nil then
        self["参数"]["on结束"](_____539F_56E0, self["参数"]["变量"])
    end
    return true
end
____exports["创建主动技能流程生命周期"] = function(_____53C2_6570)
    local _____5B9E_4F8B = __TS__New(_____4E3B_52A8_6280_80FD_6D41_7A0B_751F_547D_5468_671F_5B9E_73B0, _____53C2_6570)
    return {
        ["停止"] = function(_____539F_56E0)
            return _____5B9E_4F8B["停止"](_____5B9E_4F8B, _____539F_56E0)
        end,
        ["结束"] = function(_____539F_56E0)
            return _____5B9E_4F8B["结束"](_____5B9E_4F8B, _____539F_56E0)
        end,
        ["完成"] = function()
            return _____5B9E_4F8B["完成"](_____5B9E_4F8B)
        end,
        ["是否结束"] = function()
            return _____5B9E_4F8B["是否结束"](_____5B9E_4F8B)
        end,
        ["读取结束原因"] = function()
            return _____5B9E_4F8B["读取结束原因"](_____5B9E_4F8B)
        end
    }
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/00．技能模板/04．主动技能流程模板/00．流程生命周期.ts
/** @noSelfInFile */

import type { 技能阶段链结束原因 } from "../01．多阶段技能编排/06．技能阶段链执行器";
import type { 机制清理篮子 } from "../../04．机制组件/06．机制清理/01．机制清理篮子";

const jass = require("jass.common") as any;

const IsUnitType = jass.IsUnitType as (unit: any, unitType: any) => boolean;
const UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD as any;

const { registerDeathListener } = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心") as {
  registerDeathListener: (this: void, callback: (this: void, dyingUnit: any, killingUnit: any) => void) => void;
};

export type 主动技能流程结束原因 = 技能阶段链结束原因 | "目标失效" | "清理";

export interface 主动技能流程生命周期参数 {
  名称: string;
  施法者: any;
  目标?: any;
  清理?: 机制清理篮子;
  施法者死亡时取消?: boolean;
  目标死亡时取消?: boolean;
  变量?: any;
  on停止?: (this: void, 原因: 主动技能流程结束原因, 变量?: any) => void;
  on结束?: (this: void, 原因: 主动技能流程结束原因, 变量?: any) => void;
}

export interface 主动技能流程控制器 {
  停止(this: void, 原因?: 主动技能流程结束原因): boolean;
  结束(this: void, 原因: 主动技能流程结束原因): boolean;
  完成(this: void): boolean;
  是否结束(this: void): boolean;
  读取结束原因(this: void): 主动技能流程结束原因 | undefined;
}

const 活跃流程列表: 主动技能流程生命周期实现[] = [];
let 已注册死亡监听 = false;

function 单位有效(this: void, unit: any): boolean {
  return unit != null && unit !== 0 && IsUnitType(unit, UNIT_TYPE_DEAD) !== true;
}

function 移除活跃流程(this: void, 实例: 主动技能流程生命周期实现): void {
  for (let i = 活跃流程列表.length - 1; i >= 0; i--) {
    if (活跃流程列表[i] === 实例) {
      活跃流程列表.splice(i, 1);
      break;
    }
  }
}

function 确保死亡监听(this: void): void {
  if (已注册死亡监听) return;
  已注册死亡监听 = true;
  registerDeathListener(on主动技能流程单位死亡);
}

function on主动技能流程单位死亡(this: void, dyingUnit: any): void {
  for (let i = 活跃流程列表.length - 1; i >= 0; i--) {
    const 实例 = 活跃流程列表[i];
    if (实例 != null) 实例.处理单位死亡(dyingUnit);
  }
}

function on主动技能流程清理(this: void, variable?: any): void {
  const 实例 = variable as 主动技能流程生命周期实现 | undefined;
  if (实例 != null) 实例.停止("清理");
}

class 主动技能流程生命周期实现 {
  private 参数: 主动技能流程生命周期参数;
  private 已结束值 = false;
  private 结束原因?: 主动技能流程结束原因;

  constructor(参数: 主动技能流程生命周期参数) {
    this.参数 = 参数;

    if (参数.清理?.已清理()) {
      this.结束内部("清理", false);
      return;
    }
    if (参数.施法者死亡时取消 !== false && !单位有效(参数.施法者)) {
      this.结束内部("死亡", false);
      return;
    }
    if (参数.目标死亡时取消 === true && 参数.目标 != null && 参数.目标 !== 0 && !单位有效(参数.目标)) {
      this.结束内部("目标失效", false);
      return;
    }

    活跃流程列表.push(this);
    确保死亡监听();
    if (参数.清理 != null) {
      参数.清理.登记清理(参数.名称 + "-停止流程", on主动技能流程清理, this);
    }
  }

  停止(原因: 主动技能流程结束原因 = "中断"): boolean {
    return this.结束内部(原因, true);
  }

  结束(原因: 主动技能流程结束原因): boolean {
    return this.结束内部(原因, false);
  }

  完成(): boolean {
    return this.结束内部("完成", false);
  }

  是否结束(): boolean {
    return this.已结束值;
  }

  读取结束原因(): 主动技能流程结束原因 | undefined {
    return this.结束原因;
  }

  处理单位死亡(dyingUnit: any): void {
    if (this.已结束值) return;
    if (this.参数.施法者死亡时取消 !== false && dyingUnit === this.参数.施法者) {
      this.停止("死亡");
      return;
    }
    if (this.参数.目标死亡时取消 === true && dyingUnit === this.参数.目标) {
      this.停止("目标失效");
    }
  }

  private 结束内部(原因: 主动技能流程结束原因, 调用停止回调: boolean): boolean {
    if (this.已结束值) return false;
    this.已结束值 = true;
    this.结束原因 = 原因;
    移除活跃流程(this);

    if (调用停止回调 && this.参数.on停止 != null) {
      this.参数.on停止(原因, this.参数.变量);
    }
    if (this.参数.on结束 != null) {
      this.参数.on结束(原因, this.参数.变量);
    }
    return true;
  }
}

export function 创建主动技能流程生命周期(
  this: void,
  参数: 主动技能流程生命周期参数,
): 主动技能流程控制器 {
  const 实例 = new 主动技能流程生命周期实现(参数);
  return {
    停止: function 主动技能流程控制器停止(this: void, 原因?: 主动技能流程结束原因): boolean {
      return 实例.停止(原因);
    },
    结束: function 主动技能流程控制器结束(this: void, 原因: 主动技能流程结束原因): boolean {
      return 实例.结束(原因);
    },
    完成: function 主动技能流程控制器完成(this: void): boolean {
      return 实例.完成();
    },
    是否结束: function 主动技能流程控制器是否结束(this: void): boolean {
      return 实例.是否结束();
    },
    读取结束原因: function 主动技能流程控制器读取结束原因(this: void): 主动技能流程结束原因 | undefined {
      return 实例.读取结束原因();
    },
  };
}

export {};
